### DISCUSSION TITLE
How SBT search for dependencies for com.typesafe.play:sbt-plugin

### DISCUSSION THREAD

**Discussion by ppllaxxa** [playframework/playframework#11629] [Q&A] [answered]
Hello, I am curious how SBT understand which dependencies should be downloaded for com.typesafe.play:sbt-plugin.
Per my understanding by adding addSbtPlugin("com.typesafe.play" % "sbt-plugin" % playVersion) in plugins file for SBT project we instruct SBT to download play sbt plugin and all downstrem dependencies (e.g. https://scala.jfrog.io/ui/native/sbt-plugin-releases/com.typesafe.play/sbt-plugin/scala_2.12/sbt_1.0/2.7.9/). But in ivy.xml I could see only 2 dependnecies related to playframework:
<dependency org="com.typesafe.play" name="sbt-routes-compiler_2.12" rev="2.7.9" conf="compile->default(compile)"> </dependency>
    <dependency org="com.typesafe.play" name="run-support_2.12" rev="2.7.9" conf="compile->default(compile)"> </dependency>

Could someone help to understand where SBT takes info that it should download all other playframework dependencies?

**Comment by mkurz** [comment, marked as answer]
Lets put aside Play for a moment an lets talk about a plain sbt project: When you want to add a dependency to a sbt project you do that by adding it to the `build.sbt` file. For example if you want to the `slf4j-api` dependency to your project you do that by:
```
// in build.sbt
libraryDependencies += "org.slf4j" % "slf4j-api" % "2.0.6"
```

Now back to Play. Here, like you are correctly said, you add a sbt plugin to your `project/plugins.sbt` file. What you didn't mention is that you have to enable the play plugin in the `build.sbt` as well doing e.g.:
```scala
lazy val root = (project in file(".")).enablePlugins(PlayScala) // or PlayJava
```
This enabled plugin now modifies your project settings by e.g. adding dependencies to your `libraryDependencies` list. Lets take a deeper look at the `PlayScala` plugin I just mentioned in the code snippet above:
https://github.com/playframework/playframework/blob/9b8604643bb428d9eb389410f176bb0e69f74a55/dev-mode/sbt-plugin/src/main/scala/play/sbt/Play.scala#L89-L92

As you can see it requires another plugin to be loaded: `PlayWeb` (so you see sbtplugins can depend on other sbt plugins):
https://github.com/playframework/playframework/blob/9b8604643bb428d9eb389410f176bb0e69f74a55/dev-mode/sbt-plugin/src/main/scala/play/sbt/Play.scala#L41-L44
`PlayWeb` now defined `projectSettings` which are the settings that will be applied onto your project!
And this is how these settings look like:
https://github.com/playframework/playframework/blob/9b8604643bb428d9eb389410f176bb0e69f74a55/dev-mode/sbt-plugin/src/main/scala/play/sbt/PlaySettings.scala#L237-L271

These settings are applied to your project like if you would write it yourself into the `build.sbt` file.
In the example above there is no `libraryDependencies` settings, but e.g. the `PlayService` plugin you see in the snippet above does have one:
https://github.com/playframework/playframework/blob/9b8604643bb428d9eb389410f176bb0e69f74a55/dev-mode/sbt-plugin/src/main/scala/play/sbt/PlaySettings.scala#L102-L108

So the sbt plugin does the job of configuring the project for you, including setting `libraryDependencies`.

I hope that explains what you are asking for.

**Reply by ppllaxxa** [reply]
Wow, now it is absolutely clear. Thank you a lot!
